Xplat/src/main/java/vazkii/botania/common/config/ConfigDataManager.java

```diff
@@ -0,0 +1,118 @@
+import net.minecraft.server.packs.resources.ResourceManager;
+import net.minecraft.server.packs.resources.SimpleJsonResourceReloadListener;
+import net.minecraft.util.profiling.ProfilerFiller;
+
+import org.jetbrains.annotations.NotNull;
+
+import vazkii.botania.api.BotaniaAPI;
+import vazkii.botania.api.configdata.LooniumStructureConfiguration;
+import vazkii.botania.xplat.XplatAbstractions;
+
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+import static vazkii.botania.common.lib.ResourceLocationHelper.prefix;
+
+public class ConfigDataManager implements PreparableReloadListener {
+	public static void registerListener() {
+		XplatAbstractions.INSTANCE.registerReloadListener(PackType.SERVER_DATA, prefix("configdata"), new ConfigDataManager());
+	}
+
+	private final Map<ResourceLocation, LooniumStructureConfiguration> looniumConfigs = new HashMap<>();
+
+	public LooniumStructureConfiguration getEffectiveLooniumStructureConfiguration(ResourceLocation id) {
+		LooniumStructureConfiguration configuration = this.looniumConfigs.get(id);
+		return configuration != null ? configuration.getEffectiveConfig(looniumConfigs::get) : null;
+	}
+
+	private static void validateLooniumConfig(Map<ResourceLocation, LooniumStructureConfiguration> map) {
+		Set<ResourceLocation> errorEntries = new HashSet<>();
+		Set<ResourceLocation> visitedEntries = new LinkedHashSet<>();
+		do {
+			errorEntries.clear();
+			for (var entry : map.entrySet()) {
+				ResourceLocation id = entry.getKey();
+				ResourceLocation parent = entry.getValue().parent;
+				if (id.equals(parent)) {
+					BotaniaAPI.LOGGER.warn("Ignoring Loonium structure configuration, because it specified itself as parent: {}", id);
+					errorEntries.add(id);
+				} else {
+					visitedEntries.clear();
+					if (!findTopmostParent(map, id, parent, visitedEntries)) {
+						BotaniaAPI.LOGGER.warn("Ignoring Loonium structure configuration(s) without top-most parent: {}", visitedEntries);
+						errorEntries.addAll(visitedEntries);
+						break;
+					}
+				}
+			}
+			errorEntries.forEach(map::remove);
+		} while (!errorEntries.isEmpty() && !map.isEmpty());
+
+		if (!map.containsKey(LooniumStructureConfiguration.DEFAULT_CONFIG_ID)) {
+			BotaniaAPI.LOGGER.error("Default Loonium configuration not found!");
+		}
+	}
+
+	private static boolean findTopmostParent(Map<ResourceLocation, LooniumStructureConfiguration> map,
+			ResourceLocation id, ResourceLocation parent, Set<ResourceLocation> visitedEntries) {
+		if (!visitedEntries.add(id)) {
+			BotaniaAPI.LOGGER.warn("Cyclic dependency between Loonium structure configurations detected: {}", visitedEntries);
+			return false;
+		}
+		if (parent == null) {
+			return true;
+		}
+		var parentConfig = map.get(parent);
+		return parentConfig != null && findTopmostParent(map, parent, parentConfig.parent, visitedEntries);
```

Comment:
From what i can see, specifying a missing ID as the parent, will lead to the `Ignoring Loonium structure configuration(s) without top-most parent` error message. Or is that caught earlier? If not, maybe we should have an error specifically for that also